src/main/java/com/google/cloud/genomics/dataflow/readers/bam/Sharder.java

```diff
@@ -0,0 +1,204 @@
+    reader = r.reader;
+    indexStream = r.index;
+    header = reader.getFileHeader();
+    hasIndex = reader.hasIndex() && reader.indexing().hasBrowseableIndex();
+    LOG.info("Has index = " + hasIndex);
+    if (hasIndex) {
+      index = new BAMFileIndexImpl(
+          IOUtil.maybeBufferedSeekableStream(indexStream),header.getSequenceDictionary());
+    } else {
+      index = null;
+    }
+  }
+  
+  void processHeader() {
+    contigsByReference = Maps.newHashMap();
+    for (Contig contig : requestedContigs) {
+      contigsByReference.put(contig.referenceName != null ? contig.referenceName : "", contig);
+    }
+    if (contigsByReference.size() == 0 || contigsByReference.containsKey("*")) {
+      LOG.info("Outputting unmapped reads shard ");
+      c.output(new BAMShard(filePath, null, new Contig("*", 0, -1)));
+    }
+    final boolean allReferences =
+        contigsByReference.size() == 0 || contigsByReference.containsKey("");
+    LOG.info("All references = " + allReferences);
+    LOG.info("BAM has index = " + reader.hasIndex());
+    LOG.info("BAM has browseable index = " + reader.indexing().hasBrowseableIndex());
+    LOG.info("Class for index = " + reader.indexing().getIndex().getClass().getName());  
+  }
+  
+  Contig desiredContigForReference(SAMSequenceRecord reference) {
+    Contig contig = contigsByReference.get(reference.getSequenceName());
+    if (contig == null) {
+      if (allReferences) {
+        contig = new Contig(reference.getSequenceName(), 0, -1);
+      } else {
+        LOG.fine("No contig requested for reference " + reference.getSequenceName());
+        return null;
+      }
+    }
+    assert contig != null;
+    if (contig.end <= 0) {
+      contig = new Contig(contig.referenceName, contig.start, reference.getSequenceLength());
+      LOG.info("Updated length for contig for " + contig.referenceName + " to " + contig.end);
+    }
+    return contig;
+  }
+  
+  void createShardsForReference(SAMSequenceRecord reference, Contig contig) {
+    final BitSet overlappingBins = GenomicIndexUtil.regionToBins(
+        (int) contig.start, (int) contig.end);
+    if (overlappingBins == null) {
+      LOG.warning("Now overlapping bins for " + contig.start + ":" + contig.end);
+      return;
+    }
+    
+    BAMShard currentShard = null;
+    for (int binIndex = overlappingBins.nextSetBit(0); binIndex >= 0; binIndex = overlappingBins.nextSetBit(binIndex + 1)) {
+      final Bin bin = index.getBinData(reference.getSequenceIndex(), binIndex);
+      if (bin == null) {
+        continue;
+      }
+      if (LOG.isLoggable(Level.FINE)) {
+        LOG.fine("Processing bin " + index.getFirstLocusInBin(bin) + "-"
+            + index.getLastLocusInBin(bin));
+      }
+      
+      if (index.getLevelForBin(bin) != (GenomicIndexUtil.LEVEL_STARTS.length - 1)) {
+        if (LOG.isLoggable(Level.FINEST)) {
+          LOG.finest("Skipping - not lowest");
+        }
+        continue; // Skip non-lowest level bins
+      }
```

Comment:
@droazen, Depends on the length - with shorter you can go to lower levels.  It's basically an R-tree.  Here are a few references that describe this data structure of indexing BAM files:

##### A link describing the binning system and levels:

http://genomewiki.ucsc.edu/index.php/Bin_indexing_system

##### A figure from the paper by Kent et al  (2002) to describe the approach:

http://genome.cshlp.org/content/12/6/996/F7.expansion.html

##### The binning system described in SAM specs:

The UCSC binning scheme was suggested by Richard Durbin and Lincoln Stein and is explained in Kent et al. 20 In this scheme, each bin represents a contiguous genomic region which is either fully contained in or non-overlapping with another bin; each alignment is associated with a bin which represents the smallest region containing the entire alignment. The binning scheme is essentially a representation of R-tree. A distinct bin uniquely corresponds to a distinct internal node in a R-tree.

Bin A is a child of Bin B if the region represented by A is contained in B.
To find the alignments that overlap a specified region, we need to get the bins that overlap the region, and then test each alignment in the bins to check overlap. To quickly find alignments associated with a specified bin, we can keep in the index the start file offsets of chunks of alignments which all have the bin. As alignments are sorted by the leftmost coordinates, alignments having the same bin tend to be clustered together on the disk and therefore usually a bin is only associated with a few chunks.
Traversing all the alignments having the same bin usually needs a few seek calls. Given the set of bins that overlap the specified region, we can visit alignments in the order of their leftmost coordinates and stop seeking the rest when an alignment falls outside the required region. This strategy saves half of the seek calls in average.

In the BAI format, each bin may span 2^29 , 2^26 , 2^23 , 2^20 , 2^17 or 2^14 bp. Bin 0 spans a 512Mbp region, bins 1–8 span 64Mbp, 9–72 8Mbp, 73–584 1Mbp, 585–4680 128Kbp, and bins 4681–37448 span 16Kbp regions. This implies that this index format does not support reference chromosome sequences longer than 2^29 − 1.
The CSI format generalises the sizes of the bins, and supports reference sequences of the same
length as are supported by SAM and BAM.

To see the SAM spec directly, below is a link:

http://samtools.github.io/hts-specs/SAMv1.pdf

Hope it helps,
Paul